stores/tesco.js

```diff
@@ -0,0 +1,198 @@
+const parser = require("node-html-parser");
+const utils = require("./utils");
+
+const units = {
+    kus: { unit: "kus", factor: 1 },
+};
+
+exports.getCanonical = function (item, today) {
+    return utils.convertUnit(
+        {
+            id: "" + item.id,
+            name: item.name,
+            // description: "", not available
+            price: item.price,
+            priceHistory: [{ date: today, price: item.price }],
+            unit: item.unit,
+            quantity: item.quantity,
+            categoryNames: item.category,
+            ...(item.name && /( bio )|(^bio )/i.test(item.name) && { bio: true }),
+        },
+        units,
+        "tesco"
+    );
+};
+
+function roundNum(num, decimalExp = [1000, 100]) {
+    if (num > 1) return Math.round(num * decimalExp[1]) / decimalExp[1];
+    return Math.round(num * decimalExp[0]) / decimalExp[0];
+}
+function readPrice(txt) {
+    let real = txt
+        .match(/([ 0-9,.]+) \S+$/)[1]
+        .replace(/,/, ".")
+        .replace(/ /, "");
+    let num = parseFloat(real);
+    if (isNaN(num)) {
+        console.log(`Tesco ${num} invalid.`);
+        return real;
+    }
+    return num;
+}
+exports.fetchData = async function () {
+    let headers = { cookie: [] };
+    async function firstGet(fetchOpts, Url) {
+        let res = undefined;
+        await fetch(Url, fetchOpts).then((response) => {
+            res = response;
+            for (const pair of response.headers) {
+                console.log(`all0 ${pair[0]}:${pair[1]}`);
+                if (pair[0] == "content-length") continue;
+                if (pair[0] == "set-cookie") {
+                    headers.cookie.push(pair[1].split(";")[0]);
+                } else {
+                    headers[pair[0]] = pair[1];
+                }
+            }
+        });
+        fetchOpts.headers.cookie = headers.cookie.join("; ");
+        headers.cookie = [];
+        txt = await res.text();
+        let magics = parser.parse(txt.substring(txt.indexOf("<html")), parserOpts);
+        let scripts = magics.getElementsByTagName("script");
+        let body = scripts
+            .pop()
+            .innerText.match(/(\{"bm-verify":.+:\s+)j\}\)/)
+            .pop();
+        let mth = scripts.pop().innerText.match(/(\d+)/g);
+        let j = Number(mth[0]) + Number(mth[1] + mth[2]);
+        body += j + "}";
```

Comment:
<details>
<summary>Nejdřív dorazí ověřovací stránka, pak se nastaví cookies a třeba bez ak_bmsc to vůbec nejelo resp. snad by stačila.</summary>

```
<!DOCTYPE html>
<html>

<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta http-equiv="refresh"
        content="50; URL='/groceries/cs-CZ/shop/ovoce-a-zelenina/all?page=1&bm-verify=A...'" />
    <title>&nbsp;</title>
    <script>        var i = 1704364239;
        var j = i + Number("4469" + "83691");
    </script>
</head> <noscript> <iframe style="border: none; height: 100%; width: 100%;" src=""></iframe> </noscript>
<script> var xhr = new XMLHttpRequest();
    xhr.withCredentials = true;
    xhr.addEventListener("loadend", function () {
        try {
            var data = JSON.parse(xhr.responseText);
            if (data.hasOwnProperty('reload')) {
                if (data["reload"] == true) {
                    window.location.replace(window.location.href.replace(/[&?]bm-verify=[^#]*/, ""));
                    if (window.location.hash) {
                        window.location.reload();
                    }
                }
            } else if (data.hasOwnProperty('location')) {
                window.location.replace(data["location"]);
            } else {
                window.location.reload();
            }
        } catch (e) {
            var data = {};
            window.location.reload();
        }
    });
    xhr.open("POST", "/_sec/verify?provider=interstitial", false);
    xhr.setRequestHeader("Content-Type", "application/json");
    xhr.send(JSON.stringify({
        "bm-verify": "A...==",
        "pow": j
    }));
</script>
</body>

</html>
```
</details>